lib/workers/repository/process/vulnerabilities.ts

```diff
@@ -0,0 +1,142 @@
+import { getManagerConfig, mergeChildConfig } from '../../../config';
+import type { PackageRule, RenovateConfig } from '../../../config/types';
+import { logger } from '../../../logger';
+import type {
+  PackageDependency,
+  PackageFile,
+} from '../../../modules/manager/types';
+
+export class Vulnerabilities {
+  private osvOffline: OsvOffline | undefined;
+
+  private static readonly managerEcosystemMap: Record<
+    string,
+    Ecosystem | undefined
+  > = {
+    bundler: 'RubyGems',
+    cargo: 'crates.io',
+    gomod: 'Go',
+    gradle: 'Maven',
+    maven: 'Maven',
+    meteor: 'npm',
+    npm: 'npm',
+    nuget: 'NuGet',
+    'pip-compile': 'PyPI',
+    pip_requirements: 'PyPI',
+    pip_setup: 'PyPI',
+    pipenv: 'PyPI',
+    poetry: 'PyPI',
+    'setup-cfg': 'PyPI',
+    sbt: 'Maven',
+  };
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  private constructor() {}
+
+  private async initialize(): Promise<void> {
+    this.osvOffline = await OsvOffline.create();
+  }
+
+  static async create(): Promise<Vulnerabilities> {
+    const instance = new Vulnerabilities();
+    await instance.initialize();
+    return instance;
+  }
+
+  async fetchVulnerabilities(
+    config: RenovateConfig,
+    packageFiles: Record<string, PackageFile[]>
+  ): Promise<void> {
+    const managers = Object.keys(packageFiles).filter(
+      (manager) => Vulnerabilities.managerEcosystemMap[manager] !== undefined
+    );
+    const allManagerJobs = managers.map((manager) =>
+      this.fetchManagerVulnerabilities(config, packageFiles, manager)
+    );
+    await Promise.all(allManagerJobs);
+  }
+
+  private async fetchManagerVulnerabilities(
+    config: RenovateConfig,
+    packageFiles: Record<string, PackageFile[]>,
+    manager: string
+  ): Promise<void> {
+    const managerConfig = getManagerConfig(config, manager);
+    const queue = packageFiles[manager].map(
+      (pFile) => (): Promise<void> =>
+        this.fetchManagerPackagerFileUpdates(config, managerConfig, pFile)
+    );
+    logger.trace(
+      { manager, queueLength: queue.length },
+      'fetchManagerUpdates starting'
+    );
+    await pAll(queue, { concurrency: 5 });
+    logger.trace({ manager }, 'fetchManagerUpdates finished');
+  }
+
+  private async fetchManagerPackagerFileUpdates(
+    config: RenovateConfig,
+    managerConfig: RenovateConfig,
+    pFile: PackageFile
+  ): Promise<void> {
+    const { packageFile } = pFile;
+    const packageFileConfig = mergeChildConfig(managerConfig, pFile);
+    const { manager } = packageFileConfig;
+    const queue = pFile.deps.map(
+      (dep) => (): Promise<PackageRule[]> =>
+        this.fetchDependencyVulnerabilities(packageFileConfig, dep)
+    );
+    logger.trace(
+      { manager, packageFile, queueLength: queue.length },
+      'fetchManagerPackagerFileUpdates starting with concurrency'
+    );
+
+    config.packageRules?.push(
+      ...(await pAll(queue, { concurrency: 5 })).flat()
```

Comment:
so we have 5x5 concurrency. not sure if it's to much?

@rarkins WDYT?

Reply:
It's the same as in fetch:
- https://github.com/renovatebot/renovate/blob/8bd77f747b3da59cc955b2fc4ba3301bcb680ea2/lib/workers/repository/process/fetch.ts#L72
- https://github.com/renovatebot/renovate/blob/8bd77f747b3da59cc955b2fc4ba3301bcb680ea2/lib/workers/repository/process/fetch.ts#L90

At least this doesn't make HTTP requests, only

Reply:
Yeah, 5 is fine